Add the meta primary

ekinanp opened the pull request.

ekinanp committed:
Remove the Boolean primary

This is more of a "theoretical" than a "practical" primary so let's go
ahead and remove it. No reason to keep it in `wash find` either.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Separate negatable predicate expressions and predicate expressions

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Address PE/NPE changes in the rest of the code

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the object and array predicates

This commit also refactors some of the other value predicate code to
conform to the new AST grammar with PEs/NPEs.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the meta primary

The schema predicate will be implemented in a separate commit

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
So making Object/Array predicate accept an `NPE ValuePredicate` gives the RQL some generality that `wash find` doesn't have (e.g. it paves the way for people to construct predicates on value _types_), but it also makes metadata schema stuff hard. Having Object/Array predicate accept `PE ValuePredicate` means we can mirror the metadata schema stuff in `find` here (by importing that key sequence stuff). But we also lose that type-checking generality.

Metadata schemas were built before the `kind` primary; with `kind`, metadata schemas aren't really necessary when it comes to optimizing find. I guess they are useful for making something like `find -m .tags[?] .key termination_date -exists` expressive if the CWD is the `instances` directory so meh (other way is to write `find -k '*instance' -m .tags[?] .key termination_date -exists` but that can be a bit verbose).

I like keeping the RQL general, but maybe that generality isn't necessary. E.g. when would you ever care about a metadata value's type? I guess if you were encoding type-values as strings (like we do for `termination_date` tag?). Even then, why not just query a specific value?

So seems like `PE ValuePredicate` seems like a good idea.

ekinanp commented:
Actually ignore my earlier comment. I think there's a way I could do the schemaP while still preserving the current grammar.

ekinanp commented:
I'm going to tack on the metadata schema stuff here.

MikaelSmith commented on api/rql/internal/predicate/boolean.go:
```diff
@@ -16,11 +19,21 @@ type boolean struct {
 func (p *boolean) Unmarshal(input interface{}) error {
-	val, ok := input.(bool)
+	if !matcher.Array(matcher.Value("boolean"))(input) {
+		return errz.MatchErrorf("must be formatted as ['boolean', <value>]")
```
I'm not really sure why the type name is needed, as JSON has boolean types.

ekinanp replied:
Keep symmetric w/ the other types (only if said type has more than one possible value). We have "string", "number", "object", "array". "boolean" has two values true/false, so it made sense to me to phrase it that way. `nil` has only one value, so that's why it's labeled as such.

I guess I can go back to `true`/`false` if that's preferred too.

MikaelSmith replied:
I think I'd like to see all of those as a set of examples before deciding. All of those have types in JSON, so I don't remember why we need to call out their types.

MikaelSmith commented on api/rql/ast/metaPrimaryRealWorld_test.go:
```diff
@@ -0,0 +1,170 @@
+		),
+	)
+
+	s.TTC("amiLaunchIndex", s.A("OR", nil, s.A("NOT", nil)))
```
So, nil or not nil? That should always pass no matter what's in the file, right?

ekinanp replied:
Yeah, this is really testing existence. I'll add a comment about that once the schema stuff's done.

MikaelSmith commented on api/rql/internal/predicate/array.go:
```diff
@@ -0,0 +1,124 @@
+
+func (p *arrayElement) Unmarshal(input interface{}) error {
+	array, ok := input.([]interface{})
+	formatErrMsg := "element predicate: must be formatted as [<element_selector>, NPE ValuePredicate]"
```
Where would they find a list of valid element selectors?

ekinanp replied:
The documentation. I do plan on putting out better error messages

MikaelSmith commented on api/rql/internal/predicate/array.go:
```diff
@@ -0,0 +1,124 @@
+			p.selector = all
+		}
+	} else {
+		firstElem, ok := array[0].(float64)
```
Why float64 if we want an integer? Is that because it was unmarshaled into `interface{}` so it did it as an untyped number?

ekinanp replied:
Oh, `json.Unmarshal` unmarshals all numbers to float64 so that's why it is here.

MikaelSmith commented on api/rql/internal/primary/meta/testdata/README.md:
```diff
@@ -0,0 +1,21 @@
+`before_munging.json` was generated by running the following type through the [jsonschema](https://github.com/alecthomas/jsonschema) library (specifically ekinanp's fork):
```
Are you working on upstreaming these?

MikaelSmith replied:
Looks like schema_test.go mentions the issue created to start that process.

ekinanp replied:
Yeah, haven't gotten around to it.

MikaelSmith commented on api/rql/internal/primary/meta/schema.go:
```diff
@@ -0,0 +1,99 @@
+	// Given a value schema like ".key1.key2 5", the schema predicate's job is to check
+	// that m['key1']['key2'] == primitive_value. To do this correctly, we need to munge s
+	// a bit before returning our schema instance. That's what this code does.
+	var mungeProperties func(map[string]*jsonschema.Type) map[string]*jsonschema.Type
```
Not technically needed because `mungeProperties` isn't recursive.

ekinanp replied:
Yes, but opted to keep it because code looks a little prettier.

MikaelSmith commented on api/rql/internal/primary/meta/valueSchema.go:
```diff
@@ -0,0 +1,101 @@
+)
+
+/*
+ValueSchema represents a metadata value's schema. It is used to generate
```
Looks like all the Schema and ValueSchema stuff in meta is not yet used.

ekinanp replied:
Yeah, my annoying habit of pushing stuff up before it is fully finished. I'll comment when the schema stuff's done.

ekinanp commented:
Still working on metadata schemas. I pushed up an update that slightly changed the metadata schema framework; the change came as a result of me trying to formalize all the semantics as can be seen in `meta/valuePredicate.go`.

ekinanp committed:
Add framework for value schema predicates

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement schema predicates for each predicate and predicate expression

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement metadata schema predicates for the meta primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Ok, this is ready for review. I split up all the schema work into their own commits.

I think I'm leaning towards changing `["boolean", <value>]` back to just `<value>` since JSON does have Boolean types. I'll wait once all the schema stuff's reviewed before I do that.

ekinanp commented on api/rql/internal/primary/meta/valuePredicate.go:
```diff
@@ -0,0 +1,98 @@
+				[["key", "foo"],
+				["array",
+					["some",
+						["NOT", ["number", ["=", 5]]]]]]
```
Ah thanks, bad copy-paste

ekinanp commented on api/rql/internal/primary/meta/valuePredicate.go:
```diff
@@ -0,0 +1,98 @@
+operators].
+
+As an example, consider the following AST:
+	["object",
```
It's an internal package so won't be documented. This is meant for maintainers.

ekinanp committed:
Address PR review feedback

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Go back to specifying Boolean values

Using ["boolean", <val>] for symmetry reasons isn't very practical since
JSON has Boolean types. So go back to how it used to be.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Updated.
